src/snowflake/cli/api/utils/definition_rendering.py

```diff
@@ -0,0 +1,191 @@
+from __future__ import annotations
+
+import copy
+import os
+from dataclasses import dataclass
+from typing import List
+
+import yaml
+from jinja2 import Environment, UndefinedError, nodes
+from snowflake.cli.api.project.schemas.project_definition import ProjectDefinition
+from snowflake.cli.api.utils.rendering import get_snowflake_cli_jinja_env
+
+
+@dataclass
+class GraphNode:
+    path: str
+    value: str | None = None
+
+
+def _get_referenced_vars(ast_node, attr_chain: List[str] = []) -> List[List[str]]:
+    # Var nodes end in Name node, and starts with Getattr node.
+    # Example: ctx.env.test will look like Getattr(test) -> Getattr(env) -> Name(ctx)
+    all_referenced_vars = []
+
+    if isinstance(ast_node, nodes.Getattr):
+        attr_chain = [getattr(ast_node, "attr")] + attr_chain
+    elif isinstance(ast_node, nodes.Name):
+        all_referenced_vars = [[getattr(ast_node, "name")] + attr_chain]
+
+    for child_node in ast_node.iter_child_nodes():
+        all_referenced_vars = all_referenced_vars + _get_referenced_vars(
+            child_node, attr_chain
+        )
+
+    return all_referenced_vars
```

Comment:
Is each var a string (e.g. `ctx.env.abc`)?

Reply:
yes

Reply:
They're not stored as strings though, they're `Variable` instances (hint hint, we need type annotations throughout)